synapse/handlers/sso.py

```diff
@@ -701,8 +707,69 @@ async def _register_mapped_user(
         await self._store.record_user_external_id(
             auth_provider_id, remote_user_id, registered_user_id
         )
+
+        # Set avatar, if available
+        if attributes.picture:
+            await self.set_avatar(registered_user_id, attributes.picture)
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        try:
+            uid = UserID.from_string(user_id)
+
+            # HEAD request to find image size & mime type before download
+            response = await self._http_client.request("HEAD", picture_https_url)
+            if response.code != 200:
+                raise Exception("error sending HEAD request to get image size")
+
+            content_length = int(
+                response.headers.getRawHeaders(b"Content-Length")[0].decode("utf-8")
+            )
+            content_type = response.headers.getRawHeaders(b"Content-Type")[0].decode(
+                "utf-8"
+            )
+
+            # ensure picture size respects max_avatar_size defined in config
+            if self._profile_handler.max_avatar_size is not None:
+                if content_length > self._profile_handler.max_avatar_size:
+                    raise Exception("sso avatar too big in size")
+
+            # ensure picture is of allowed mime type
+            if (
+                self._profile_handler.allowed_avatar_mimetypes
+                and content_type not in self._profile_handler.allowed_avatar_mimetypes
+            ):
+                raise Exception("mime type not allowed for sso avatar")
+
+            # download picture
+            response = await self._http_client.request("GET", picture_https_url)
```

Comment:
A malicious server could say that it will only send us 1000 bytes in the HEAD content length, but then proceed to send us 100000000000000 bytes in the body of the GET response. 

Is there a way to impose the size limit here too?

Reply:
I looked at the HTTP client and it has a method [`get_file`](https://github.com/matrix-org/synapse/blob/730b13dbc9e48181b1aaf38be870ec21364b1e9c/synapse/http/client.py#L704) which ensures a maximum size (retrieving the data with a function called `read_body_with_max_size`). Possibly this is a path forward?

Reply:
Thanks @akirk I was able to get that done using `get_file()` method 👍

Reply:
>> A malicious server could say that it will only send us 1000 bytes in the HEAD content length, but then proceed to send us 100000000000000 bytes in the body of the GET response.
> Aren't http clients supposed to read response body only till Content-Length?

Yes, but there's nothing stopping the server from saying `Content-Length: 100000000000000` in the GET response headers.